code/game/machinery/autolathe.dm

```diff
@@ -47,45 +46,272 @@
+				return
+			for(var/mat in materials.materials)
+				var/datum/material/M = mat
+				if("[M.type]" == material_datum)
+					materials.retrieve_sheets(amount, M, get_release_direction())
+					return
+		if("queue_repeat")
+			queue_repeating = text2num(params["repeating"])
+		if("clear_queue")
+			item_queue = list()
+		if("item_repeat")
+			var/design_id = params["design_id"]
+			var/repeating_mode = text2num(params["repeating"])
+			if(!item_queue["[design_id]"])
+				return
+			item_queue["[design_id]"]["repeating"] = repeating_mode
+		if("clear_item")
+			var/design_id = params["design_id"]
+			var/queue_type = text2num(params["build_queue"])
+			if(queue_type)
+				build_queue -= design_id
+			else
+				item_queue -= design_id
+		if("queue_item")
+			var/design_id = params["design_id"]
+			var/amount = text2num(params["amount"])
+			add_to_queue(item_queue, design_id, amount)
+		if("build_item")
+			var/design_id = params["design_id"]
+			var/amount = text2num(params["amount"])
+			add_to_queue(build_queue, design_id, amount)
+		if("begin_process")
+			begin_process()
+	//Update the UI for them so it's smooth
+	ui_interact(usr)
+
+/obj/machinery/autolathe/proc/update_viewer_statics()
+	for(var/mob/M in viewing_mobs)
+		if(QDELETED(M) || !(M.client || M.mind))
+			continue
+		update_static_data(M)
 
-	var/datum/browser/popup = new(user, "autolathe", name, 400, 500)
-	popup.set_content(dat)
-	popup.open()
+/obj/machinery/autolathe/proc/add_to_queue(queue_list, design_id, amount, repeat=0)
+	if(queue_list["[design_id]"])
+		queue_list["[design_id]"]["amount"] += amount
+		if(queue_list["[design_id]"]["amount"] <= 0)
+			queue_list -= "[design_id]"
+		return
+	if(amount <= 0)
+		return
+	//Check if the item uses custom materials
+	var/datum/design/requested_item = stored_research.isDesignResearchedID(design_id)
+	var/datum/material/used_material = null
+	for(var/MAT in requested_item.materials)
+		used_material = MAT
+		if(istext(used_material)) //This means its a category
+			var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+			var/list/list_to_show = list()
+			for(var/i in SSmaterials.materials_by_category[used_material])
+				if(materials.materials[i] > 0)
+					list_to_show += i
+			used_material = input("Choose [used_material]", "Custom Material") as null|anything in sortList(list_to_show, /proc/cmp_typepaths_asc)
+			if(!used_material)
+				return //Didn't pick any material, so you can't build shit either.
+
+	queue_list["[design_id]"] = list(
+		"amount" = amount,
+		"repeating" = repeat,
+		"build_mat" = used_material,
+	)
+
+/obj/machinery/autolathe/proc/get_release_direction()
+	. = get_turf(src)
+	switch(output_direction)
+		if("up")
+			. = get_offset_target_turf(src, 0, 1)
+		if("down")
+			. = get_offset_target_turf(src, 0, -1)
+		if("right")
+			. = get_offset_target_turf(src, 1, 0)
+		if("left")
+			. = get_offset_target_turf(src, -1, 0)
+	if(is_blocked_turf(., TRUE))
```

Comment:
Probably bad practice to use . like this. I would make a turf typed var and just return it at the end

Reply:
Changed